README.md
<p align="center">
<img height="200" src="./assets/kv.png" alt="to unocss">
</p>
<p align="center"> English | <a href="./README_zh.md">简体中文</a></p>

The vscode extension converts [unocss](https://github.com/unocss/unocss) to css.It's powered by @unocss/core.

![demo](/assets/demo.gif)

## Config
```
// Custom configuration highlighting style
// settings.json
  "unocss-to-css":{
    "dark": {
      "textDecoration": "underline",
      "backgroundColor": "rgba(144, 238, 144, 0.5)",
      "color": "black",
    },
    "light":{
      "textDecoration": "underline",
      "backgroundColor": "rgba(255, 165, 0, 0.5)",
      "color": "#ffffff",
    }
  }
```

## :coffee:

[buy me a cup of coffee](https://github.com/Simon-He95/sponsor)

## License

[MIT](./license)


CHANGELOG.md
# Change Log

All notable changes to the "unocss-to-css" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/extension.ts
import * as vscode from 'vscode'
import { transformUnocssBack } from './utils'

// 插件被激活时调用activate
export function activate() {
  const LANS = ['html', 'vue', 'swan', 'wxml', 'axml', 'css', 'wxss', 'acss', 'less', 'scss', 'sass', 'stylus', 'wxss', 'acss']

  // 注册hover事件
  vscode.languages.registerHoverProvider(LANS, {
    async provideHover(document, position) {
      // 获取当前选中的文本范围
      const editor = vscode.window.activeTextEditor
      if (!editor)
        return
      const selection = editor.selection
      const wordRange = new vscode.Range(selection.start, selection.end)
      let selectedText = editor.document.getText(wordRange)
      if (!selectedText) {
        const range = document.getWordRangeAtPosition(position)
        let word = document.getText(range)
        const lineNumber = position.line
        const line = document.lineAt(lineNumber).text
        const wholeReg = new RegExp(`(\\w+:)?([\\w\\-\\[\\(\\!]+)?${word}(:*[^"\\s\\/>]+)?`)
        const matcher = line.match(wholeReg)
        if (matcher)
          word = matcher[0]
        const equalReg = new RegExp(`([\\w\\-]+)=["'][^"']*${word}[^"']*["']`)
        const match = line.match(equalReg)
        if (match && match[1] !== 'class')
          word = `${match[1]}-${word}`
        selectedText = word
      }

      if (!selectedText)
        return
      const css = await transformUnocssBack(selectedText) as string
      if (!css)
        return
      const md = new vscode.MarkdownString()
      md.appendMarkdown(`<span style="color:green;font-weight: bold">${css}</span>\n`)

      return new vscode.Hover(md)
    },
  })
}

// this method is called when your extension is deactivated
export function deactivate() { }


src/utils/index.js
import { createGenerator } from '@unocss/core';
import presetUno from '@unocss/preset-uno';
export function transformUnocssBack(code) {
    return new Promise((resolve) => {
        createGenerator({}, {
            presets: [presetUno()],
        })
            .generate(code || '')
            .then((res) => {
            const css = res.getLayers();
            const reg = new RegExp(`${escapeRegExp(code)}([:\\>][\\w\\-\\(\\)]+)?{(.*)}`);
            const match = css.match(reg);
            if (!match)
                return resolve('');
            const result = match[0].replace(match[2], match[2].replace(/[:,]/g, v => `${v} `)).replace('{', ' {\n  ').replace(/;/g, ';\n  ').replace('  }', '}');
            resolve(result);
        });
    });
}
function escapeRegExp(str) {
    return str.replace(/[.*+?^${}()|[\]\\%:\!\&\>]/g, '\\\\\\$&');
}


src/index.ts
import * as vscode from 'vscode'
import { transformUnocssBack } from './utils'

// 插件被激活时调用activate
export function activate() {
  // 将规则添加到语言配置中
  const LANS = ['html', 'vue', 'swan', 'wxml', 'axml', 'css', 'wxss', 'acss', 'less', 'scss', 'sass', 'stylus', 'wxss', 'acss']
  const { dark, light } = vscode.workspace.getConfiguration('unocss-to-css') || {}
  const style = Object.assign({
    dark: {
      textDecoration: 'underline',
      backgroundColor: 'rgba(144, 238, 144, 0.5)',
      color: 'black',
    },
    light: {
      textDecoration: 'underline',
      backgroundColor: 'rgba(255, 165, 0, 0.5)',
      color: '#ffffff',
    }
  }, {
    dark,
    light
  })
  const decorationType = vscode.window.createTextEditorDecorationType(style)

  // 注册hover事件
  vscode.languages.registerHoverProvider(LANS, {
    async provideHover(document, position) {
      // 获取当前选中的文本范围
      const editor = vscode.window.activeTextEditor
      if (!editor)
        return
      // 移除样式
      vscode.window.activeTextEditor?.setDecorations(decorationType, [])
      const selection = editor.selection
      const wordRange = new vscode.Range(selection.start, selection.end)
      let selectedText = editor.document.getText(wordRange)
      const realRangeMap: any = []
      if (!selectedText) {
        const range = document.getWordRangeAtPosition(position) as any
        const line = range.c.c
        let word = document.getText(range)
        const lineNumber = position.line
        const lineText = document.lineAt(lineNumber).text
        const wholeReg = new RegExp(`([\\w\\>\\[]+:)?(\\[[^\\]]+\\]:)?([\\w\\-\\[\\(\\!\\>\\&]+)?${word}(:*[^"\\s\\/>]+)?`, 'g')
        let matcher = null

        for (const match of lineText.matchAll(wholeReg)) {
          const { index } = match
          const pos = index! + match[0].indexOf(word)
          if (pos === range?.c?.e) {
            matcher = match
            realRangeMap.push({
              content: match[0],
              range: new vscode.Range(
                new vscode.Position(line, index!),
                new vscode.Position(line, index! + match[0].length),
              ),
            })
            break
          }
        }
        if (matcher)
          word = matcher[0]
        matcher = null
        const equalReg = new RegExp(`([\\[\\]\\(\\)\\>\\w\\-]+)=["'][^"']*${word}[^"']*["']`, 'g')
        for (const match of lineText.matchAll(equalReg)) {
          // 找比range小但最近的
          const { index } = match
          if (index! > range?.c?.e)
            break
          matcher = match
        }
        if (matcher && matcher[1] !== 'class') {
          word = `${matcher[1]}-${word}`
          realRangeMap.push({
            range: new vscode.Range(
              new vscode.Position(line, matcher.index!),
              new vscode.Position(line, matcher.index! + matcher[1].length),
            ),
            content: matcher[1],
          })
        }
        selectedText = word
      }

      if (!selectedText)
        return
      const css = await transformUnocssBack(selectedText) as string
      if (!css)
        return

      // 增加decorationType样式
      editor.edit(() => editor.setDecorations(decorationType, realRangeMap.map((item: any) => item.range)))
      const md = new vscode.MarkdownString()
      md.isTrusted = true
      md.supportHtml = true
      md.appendMarkdown('<a href="https://github.com/Simon-He95/unocss-to-css">Unocss To Css:</a>\n')
      md.appendCodeblock(css, 'css')
      return new vscode.Hover(md)
    },
  })

  // 监听编辑器选择内容变化的事件
  vscode.window.onDidChangeTextEditorSelection((event) => {
    vscode.window.activeTextEditor?.setDecorations(decorationType, [])
  })
}

// this method is called when your extension is deactivated
export function deactivate() { }
